File: mybatis/src/main/java/data/spring/mybatis/application/service/product/command/ProductUpdateCommand.java
package data.spring.mybatis.application.service.product.command;

public record ProductUpdateCommand(
        Long productId,
        String productName,
        Integer price,
        Integer quantity
){
    public static ProductUpdateCommand of(Long productId, String productName, Integer price, Integer quantity) {
        return new ProductUpdateCommand(productId, productName, price, quantity);
    }
}

File: mybatis/src/main/kotlin/data/spring/mybatis/application/service/product/command/ProductUpdateCommand.kt
package data.spring.mybatis.application.service.product.command

import data.spring.mybatis.domain.clock
import java.time.LocalDateTime

class ProductUpdateCommand(
    val productId: Long,
    val productName: String?,
    val price: Int?,
    val quantity: Int?,
) {
    var updatedAt: LocalDateTime? = null

    fun isUpdated(): Boolean {
        return productName != null || price != null || quantity != null
    }

    fun whenUpdated(): ProductUpdateCommand {
        updatedAt = LocalDateTime.now(clock())
        return this
    }
}
